client/telemetry/src/layer.rs

```diff
@@ -0,0 +1,146 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use futures::channel::mpsc;
+use parking_lot::Mutex;
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::sync::Arc;
+use tracing::{Event, Id, Subscriber};
+use tracing_subscriber::{layer::Context, registry::LookupSpan, Layer};
+
+/// Span name used to report the telemetry.
+pub const TELEMETRY_LOG_SPAN: &str = "telemetry-logger";
+
+/// `Layer` that handles the logs for telemetries.
+#[derive(Debug)]
+pub struct TelemetryLayer(Senders);
+
+impl TelemetryLayer {
+	/// Create a new [`TelemetryLayer`] using the [`Senders`] provided in argument.
+	pub fn new(senders: Senders) -> Self {
+		Self(senders)
+	}
+}
+
+impl<S> Layer<S> for TelemetryLayer
+where
+	S: Subscriber + for<'a> LookupSpan<'a>,
+{
+	fn on_event(&self, event: &Event<'_>, ctx: Context<S>) {
+		if event.metadata().target() != TELEMETRY_LOG_SPAN {
+			return;
+		}
+
+		if let Some(span) = ctx.lookup_current() {
+			let parents = span.parents();
+
+			if let Some(span) = std::iter::once(span)
+				.chain(parents)
+				.find(|x| x.name() == TELEMETRY_LOG_SPAN)
+			{
+				let id = span.id();
+				if let Some(sender) = (self.0).0.lock().get_mut(&id) {
+					let mut attrs = TelemetryAttrs::new(id);
+					let mut vis = TelemetryAttrsVisitor(&mut attrs);
+					event.record(&mut vis);
+
+					if let TelemetryAttrs {
+						verbosity: Some(verbosity),
+						json: Some(json),
+						..
+					} = attrs
+					{
+						let _ = sender.try_send((
+							verbosity
+								.try_into()
+								.expect("telemetry log message verbosity are u8; qed"),
+							json,
+						));
+					} else {
+						panic!(
```

Comment:
I think we should log this, but not panicking.

Reply:
We can't log at this place.

But I think it should panic as it is really a bug

Reply:
Isn't this layer always used? Also for normal logging?

If we can't log, we can still write to stderr.

I just don't want to see nodes crashing because of this. Just imagine a rare code path that is almost never executed, just this one time and it brings down the whole network.

Reply:
Yeah clearly... :thinking: 

No I don't know why but any call to a `log!` (or whatever) in there just doesn't show anything.

I will log to STDERR then.

Reply:
Done